editor/import/scene_importer_mesh.cpp

```diff
@@ -190,67 +263,343 @@ void EditorSceneImporterMesh::generate_lods() {
+			merged_indices[j] = vertex_remap[indices[j]];
+		}
+
+		unsigned int merged_vertex_count = merged_vertices.size();
+		const Vector3 *merged_vertices_ptr = merged_vertices.ptr();
+		const int32_t *merged_indices_ptr = merged_indices.ptr();
+
+		{
+			const int *counts_ptr = merged_normals_counts.ptr();
+			Vector3 *merged_normals_ptrw = merged_normals.ptr();
+			for (unsigned int j = 0; j < merged_vertex_count; j++) {
+				merged_normals_ptrw[j] /= counts_ptr[j];
+			}
+		}
+
+		LocalVector<float> normal_weights;
+		normal_weights.resize(merged_vertex_count);
+		for (unsigned int j = 0; j < merged_vertex_count; j++) {
+			normal_weights[j] = 2.0; // Give some weight to normal preservation, may be worth exposing as an import setting
+		}
+
+		const float max_mesh_error = FLT_MAX; // We don't want to limit by error, just by index target
+		float scale = SurfaceTool::simplify_scale_func((const float *)merged_vertices_ptr, merged_vertex_count, sizeof(Vector3));
 		float mesh_error = 0.0f;
-		float scale = SurfaceTool::simplify_scale_func((const float *)vertices_ptr, vertex_count, sizeof(Vector3));
-		while (index_target > min_indices) {
-			Vector<int> new_indices;
-			new_indices.resize(indices.size());
-			size_t new_len = SurfaceTool::simplify_with_attrib_func((unsigned int *)new_indices.ptrw(), (const unsigned int *)indices.ptr(), indices.size(), (const float *)vertices_ptr, vertex_count, sizeof(Vector3), index_target, max_mesh_error_percentage, &mesh_error, (float *)attributes.ptrw(), normal_weights.ptrw(), attribute_count);
-			if ((int)new_len > (index_target * error_tolerance)) {
-				break;
+
+		unsigned int index_target = 12; // Start with the smallest target, 4 triangles
+		unsigned int last_index_count = 0;
+
+		int split_vertex_count = vertex_count;
+		LocalVector<Vector3> split_vertex_normals;
+		LocalVector<int> split_vertex_indices;
+		split_vertex_normals.reserve(index_count / 3);
+		split_vertex_indices.reserve(index_count / 3);
+
+		RandomPCG pcg;
+		pcg.seed(123456789); // Keep seed constant across imports
+
+		Ref<StaticRaycaster> raycaster = StaticRaycaster::create();
+		if (raycaster.is_valid()) {
+			raycaster->add_mesh(vertices, indices, 0);
+			raycaster->commit();
+		}
+
+		while (index_target < index_count) {
+			PackedInt32Array new_indices;
+			new_indices.resize(index_count);
+
+			size_t new_index_count = SurfaceTool::simplify_with_attrib_func((unsigned int *)new_indices.ptrw(), (const uint32_t *)merged_indices_ptr, index_count, (const float *)merged_vertices_ptr, merged_vertex_count, sizeof(Vector3), index_target, max_mesh_error, &mesh_error, (float *)merged_normals.ptr(), normal_weights.ptr(), 3);
+
+			if (new_index_count < last_index_count * 1.5f) {
+				index_target = index_target * 1.5f;
+				continue;
 			}
-			Surface::LOD lod;
-			lod.distance = mesh_error * scale;
-			if (Math::is_zero_approx(mesh_error)) {
+
+			if (new_index_count <= 0 || (new_index_count >= (index_count * 0.75f))) {
```

Comment:
Is the intention of this code that the LOD generation will stop once it surpasses an upper bound of 75% of the original mesh's indices?